File: Java Solutions/261_GraphValidTree.java
class Solution {
    public boolean validTree(int n, int[][] edges) {
        List<List<Integer>> adjacent = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adjacent.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            adjacent.get(edge[0]).add(edge[1]);
            adjacent.get(edge[1]).add(edge[0]);
        }
        
        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();
        queue.offer(0);
        visited.add(0);
        Map<Integer, Integer> parent = new HashMap<>();

        while (!queue.isEmpty()) {
            int node = queue.poll();
            for (int nei : adjacent.get(node)) {
                if (parent.getOrDefault(node, -1) == nei) continue;
                if (visited.contains(nei)) return false;
                parent.put(nei, node);
                queue.offer(nei);
                visited.add(nei);
            }
        }
        return visited.size() == n;
    }
}

File: Graph/261_GraphValidTree.py
# BFS
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        adjacent = defaultdict(list)
        for n1, n2 in edges:
            adjacent[n1].append(n2)
            adjacent[n2].append(n1)

        queue = deque([0])
        visited = set([0])
        parent = {}
        while queue:
            node = queue.popleft()
            for nei in adjacent[node]:
                if parent.get(node) == nei:
                    continue
                if nei in visited:
                    return False
                parent[nei] = node
                queue.append(nei)
                visited.add(nei)
        return len(visited) == n


# DFS
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        adjacent = defaultdict(list)
        for n1, n2 in edges:
            adjacent[n1].append(n2)
            adjacent[n2].append(n1)

        visited = set()

        def dfs(i, prev):
            if i in visited:
                return False
            visited.add(i)
            for j in adjacent[i]:
                if j == prev:
                    continue
                if not dfs(j, i):
                    return False
            return True

        return dfs(0, -1) and len(visited) == n


# Advanced Graph Theory + Iterative DFS
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        # For the graph to be a valid tree, it must have exactly n - 1 edges
        if len(edges) != n - 1:
            return False

        adj = defaultdict(list)
        for u, v in edges:
            adj[u].append(v)
            adj[v].append(u)

        visited = {0}
        stack = [0]

        while stack:
            node = stack.pop()
            for nei in adj[node]:
                if nei in visited:
                    continue
                visited.add(nei)
                stack.append(nei)

        return len(visited) == n